### DISCUSSION TITLE
Lifter Architecture Discussion

### DISCUSSION THREAD

**Discussion by bayedieng** [obhq/obliteration#268] [General]
In order to start the lifter I was thinking of using the methodology described in chapter 3.3 of this [paper](https://mediatum.ub.tum.de/doc/1614897/1614897.pdf). The basic idea is to take output from the disassembler and to create a function from the machine code comprised of basic blocks of non-branching instructions. Once the function is created each block will be translated to equivalent LLVM IR on-the-fly to end up with a Single LLVM [Function](https://llvm.org/doxygen/classllvm_1_1Function.html) comprised of basic blocks of LLVM IR. This function would then be used to generate target machine code through the LLVM Framework. 

This is the approach i'd like to go with but if anyone has a better proposal feel free to chime in.

**Comment by ultimaweapon** [comment]
Let me know if you need help related to the Obliteration like how to get the entry point function, etc.

**Comment by ultimaweapon** [comment]
@bayedieng are you currently working on `LlvmEngine`? We need to re-structure the execution engines after my next PR is merged.

**Comment by bayedieng** [comment]
Yes, I wanted to add a vec of basic blocks of non-branching instructions instead of the vec of instructions in the ```Function``` struct of the code generator but you can restructure it first and I'll do so afterwards.

**Comment by ultimaweapon** [comment]
Are you currently using `src/kernel/src/disasm`? If not I'll remove it.

**Reply by bayedieng** [reply]
Nope, you can go ahead and remove it.

**Comment by bayedieng** [comment]
Hey @obhq/developers, I haven't had much time to work on the lifter in the past month and probably won't be having much time to work extensively in the future either. If anyone wants to tackle binary translation before I personally get to it feel free to do so. I think the best approach would be to translate a single first (was thinking about ```mov```) then incrementally add others. 

Here are some references I've been looking at:
- [Revng: Binary Translation using Qemu's TCG IR](https://github.com/revng/revng) 
    - Qemu TCG's IR is known to be quite slow and GPL License
    - GPL License
- [Rellume: LLVM IR lifter referenced in paper in OP](https://github.com/revng/revng)
    - GPL License
- [Efficient LLVM-Based Dynamic Binary Translation](https://home.in.tum.de/~engelke/pubs/2104-vee.pdf) 
   - Seems to be fastest method using LLVM IR for Binary Translation
   - here's [code](https://github.com/aengelke/instrew) LGPL License
